cmd/complex/main.go

```diff
@@ -25,28 +26,41 @@ func newSSE() *sse.Server {
 	rp, _ := sse.NewValidReplayer(time.Minute*5, true)
 	rp.GCInterval = time.Minute
 
+	logFunc := func(r *http.Request) *slog.Logger {
+		l, err := getLogger(r)
+		if err != nil {
+			return nil
+		}
+		return l
+	}
+
 	return &sse.Server{
 		Provider: &sse.Joe{Replayer: rp},
-		Logger:   logger{log.New(os.Stderr, "", 0)},
-		OnSession: func(s *sse.Session) (sse.Subscription, bool) {
-			topics := s.Req.URL.Query()["topic"]
+		Logger:   logFunc,
+		OnSession: func(w http.ResponseWriter, r *http.Request) (topics []string, permitted bool) {
+			topics = r.URL.Query()["topic"]
 			for _, topic := range topics {
 				if topic != topicRandomNumbers && topic != topicMetrics {
-					fmt.Fprintf(s.Res, "invalid topic %q; supported are %q, %q", topic, topicRandomNumbers, topicMetrics)
-					s.Res.WriteHeader(http.StatusBadRequest)
-					return sse.Subscription{}, false
+					fmt.Fprintf(
+						w,
+						"invalid topic %q; supported are %q, %q",
+						topic,
+						topicRandomNumbers,
+						topicMetrics,
+					)
+
+					// NOTE: if you are returning false to reject the subscription you must also return
+					// your own error code!
+					w.WriteHeader(http.StatusBadRequest)
```

Comment:
Technically this is not true – if a `ServeHTTP` ends without a response code, a `200 OK` will be implicitly written.

Let's maybe change this to something like:
```suggestion
					// NOTE: if you are returning false to reject the subscription we recommend
					// writing your own response code. Clients will receive a confusing 200 OK otherwise.
					w.WriteHeader(http.StatusBadRequest)
```

Reply:
✅